File: src/orm/validators.ts
import type { ColumnOptions, ValidatorFn } from './data-types';

/**
 * Nomes de validadores built-in no objeto `validate` (padrão próximo do Sequelize 6),
 * além de chaves cujo valor seja `ValidatorFn`.
 */
export const BUILTIN_VALIDATOR_NAMES = new Set(
    [
        'notEmpty',
        'isNull',
        'isIn',
        'notIn',
        'min',
        'max',
        'len',
        'isEmail',
        'isUrl',
        'isIP',
        'isUUID',
        'isDate',
        'isInt',
        'isFloat',
        'isDecimal',
        'isNumeric',
        'isAlpha',
        'isAlphanumeric',
        'matches',
        'contains',
        'notContains',
        'isAfter',
        'isBefore',
        'equals',
        'not'
    ].map((s) => s.toLowerCase())
);

export function isBuiltInValidatorKey(name: string): boolean {
    return BUILTIN_VALIDATOR_NAMES.has(name.toLowerCase());
}

type MsgSpec = { msg?: string; args?: any };

function unpack(spec: unknown): { msg?: string; value: any } {
    if (spec != null && typeof spec === 'object' && !Array.isArray(spec) && ('args' in spec || 'msg' in spec)) {
        const s = spec as MsgSpec;
        if ('args' in s) return { msg: s.msg, value: s.args };
    }
    return { value: spec };
}

const RE_EMAIL = /^[a-z0-9._%+\-]+@[a-z0-9.\-]+\.[a-z]{2,63}$/i;
const RE_INT = /^-?\d+$/;
const RE_NUMERIC = /^-?\d+(\.\d+)?([eE][+\-]?\d+)?$/;
const RE_ALPHA = /^[A-Za-zÀ-ÿ\u0100-\uFFFF]+$/;
const RE_ALNUM = /^[0-9A-Za-zÀ-ÿ\u0100-\uFFFF]+$/;
const RE_IPV4 = /^(?:(?:25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)\.){3}(?:25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)$/;
const RE_UUID = /^[0-9a-f]{8}-[0-9a-f]{4}-[1-5][0-9a-f]{3}-[89ab][0-9a-f]{3}-[0-9a-f]{12}$/i;
const RE_UUID4 = /^[0-9a-f]{8}-[0-9a-f]{4}-4[0-9a-f]{3}-[89ab][0-9a-f]{3}-[0-9a-f]{12}$/i;

function toDate(x: any): Date | null {
    if (x instanceof Date) return isNaN(x.getTime()) ? null : x;
    if (x == null) return null;
    const d = new Date(x);
    return isNaN(d.getTime()) ? null : d;
}

function isBlankValue(v: any): boolean {
    if (v == null) return true;
    if (typeof v === 'string' && v.trim() === '') return true;
    if (Array.isArray(v) && v.length === 0) return true;
    return false;
}

/**
 * Sequelize: `isIn: [['A','B']]` ou `isIn: ['A', 'B']`.
 */
function parseAllowedList(v: any): any[] {
    if (v == null) return [];
    if (Array.isArray(v) && v.length > 0 && Array.isArray((v as any[])[0])) {
        return (v as any[])[0] as any[];
    }
    if (Array.isArray(v)) {
        return v;
    }
    return [v];
}

/**
 * Cria `ValidatorFn` a partir de um nome e spec (ex.: `true`, `5`, `[[a,b]]`, `{ args: [5], msg: '...' }`).
 */
export function buildBuiltinValidator(validatorName: string, spec: unknown, fieldKey: string): ValidatorFn {
    const name = validatorName.toLowerCase();
    const { msg, value: arg } = unpack(spec);

    if (name === 'not') {
        return (v: any) => runNotValidator(spec, v, fieldKey, msg);
    }

    return (value: any) => {
        if (name === 'notempty') {
            if (isBlankValue(value)) return (msg as string) || `${fieldKey} cannot be empty.`;
            return true;
        }
        if (name === 'isnull') {
            if (value !== null && value !== undefined) {
                return (msg as string) || `${fieldKey} must be null.`;
            }
            return true;
        }
        if (name === 'isin' || name === 'notin') {
            if (value === null || value === undefined) return true;
            const list = parseAllowedList(arg);
            const hit = list.some((x) => Object.is(x, value) || x === value);
            if (name === 'isin' && !hit) {
                return (msg as string) || `${fieldKey} is not a valid choice.`;
            }
            if (name === 'notin' && hit) {
                return (msg as string) || `${fieldKey} is not an allowed value.`;
            }
            return true;
        }
        if (name === 'min' || name === 'max') {
            if (value === null || value === undefined) {
                return true;
            }
            const n = Number(arg);
            if (name === 'min') {
                if (typeof value === 'number' && !isNaN(value) && value < n) {
                    return (msg as string) || `${fieldKey} must be >= ${n}.`;
                }
                if (typeof value === 'string' && value.length < n) {
                    return (msg as string) || `${fieldKey} is too short (min ${n}).`;
                }
                if (Array.isArray(value) && value.length < n) {
                    return (msg as string) || `${fieldKey} has too few elements.`;
                }
                if (value instanceof Date) {
                    const cmp = toDate(n);
                    if (cmp && value < cmp) {
                        return (msg as string) || `${fieldKey} is too early.`;
                    }
                }
            } else {
                if (typeof value === 'number' && !isNaN(value) && value > n) {
                    return (msg as string) || `${fieldKey} must be <= ${n}.`;
                }
                if (typeof value === 'string' && value.length > n) {
                    return (msg as string) || `${fieldKey} is too long (max ${n}).`;
                }
                if (Array.isArray(value) && value.length > n) {
                    return (msg as string) || `${fieldKey} has too many elements.`;
                }
                if (value instanceof Date) {
                    const cmp = toDate(n);
                    if (cmp && value > cmp) {
                        return (msg as string) || `${fieldKey} is too late.`;
                    }
                }
            }
            return true;
        }
        if (name === 'len') {
            if (value === null || value === undefined) {
                return true;
            }
            const pair =
                (Array.isArray(arg) && arg.length >= 2
                    ? arg
                    : (arg as any)?.args && Array.isArray((arg as any).args) && (arg as any).args.length >= 2
                      ? (arg as any).args
                      : null) as [number, number] | null;
            if (pair) {
                const a = Number(pair[0]);
                const b = Number(pair[1]);
                if (typeof value === 'string' || Array.isArray(value)) {
                    const l = (value as any).length;
                    if (l < a || l > b) {
                        return (
                            (msg as string) ||
                            `${fieldKey} must have between ${a} and ${b} ${
                                typeof value === 'string' ? 'characters' : 'items'
                            }.`
                        );
                    }
                }
            }
            return true;
        }
        if (value === null || value === undefined) {
            return true;
        }
        if (name === 'isemail' && (arg === true || arg == null)) {
            if (!RE_EMAIL.test(String(value))) {
                return (msg as string) || `${fieldKey} must be a valid email.`;
            }
            return true;
        }
        if (name === 'isurl' && (arg === true || arg == null || (typeof arg === 'object' && arg !== null))) {
            try {
                new URL(String(value));
            } catch {
                return (msg as string) || `${fieldKey} must be a valid URL.`;
            }
            if (arg && typeof arg === 'object' && (arg as any).requireTld) {
                const h = (() => {
                    try {
                        return new URL(String(value)).hostname;
                    } catch {
                        return '';
                    }
                })();
                if (!h || !h.includes('.')) {
                    return (msg as string) || `${fieldKey} must have a TLD.`;
                }
            }
            return true;
        }
        if (name === 'isip' && (arg === true || arg == null)) {
            if (!RE_IPV4.test(String(value))) {
                return (msg as string) || `${fieldKey} must be a valid IPv4.`;
            }
            return true;
        }
        if (name === 'isuuid') {
            const s = String(value);
            if (arg === 4) {
                if (!RE_UUID4.test(s)) {
                    return (msg as string) || `${fieldKey} must be a valid UUID v4.`;
                }
            } else {
                if (!RE_UUID.test(s)) {
                    return (msg as string) || `${fieldKey} must be a valid UUID.`;
                }
            }
            return true;
        }
        if (name === 'isdate' && (arg === true || arg == null)) {
            if (!toDate(value)) {
                return (msg as string) || `${fieldKey} must be a valid date.`;
            }
            return true;
        }
        if (name === 'isint') {
            if (arg && typeof arg === 'object' && (arg as any).min != null) {
                const n = Number(value);
                if (!Number.isInteger(n)) {
                    return (msg as string) || `${fieldKey} must be an integer.`;
                }
                if (n < (arg as any).min || n > (arg as any).max) {
                    return (msg as string) || `${fieldKey} is out of range.`;
                }
            } else {
                if (!RE_INT.test(String(value)) || !Number.isInteger(Number(value))) {
                    return (msg as string) || `${fieldKey} must be an integer.`;
                }
            }
            return true;
        }
        if (name === 'isfloat' || name === 'isdecimal') {
            if (!RE_NUMERIC.test(String(value)) || isNaN(Number(value))) {
                return (msg as string) || `${fieldKey} must be a number.`;
            }
            if (name === 'isfloat' && !Number.isFinite(Number(value))) {
                return (msg as string) || `${fieldKey} must be a finite number.`;
            }
            return true;
        }
        if (name === 'isnumeric') {
            if (!RE_NUMERIC.test(String(value))) {
                return (msg as string) || `${fieldKey} must be numeric.`;
            }
            return true;
        }
        if (name === 'isalpha' && (arg === true || arg == null)) {
            if (!RE_ALPHA.test(String(value))) {
                return (msg as string) || `${fieldKey} may only contain letters.`;
            }
            return true;
        }
        if (name === 'isalphanumeric' && (arg === true || arg == null)) {
            if (!RE_ALNUM.test(String(value))) {
                return (msg as string) || `${fieldKey} may only contain letters and numbers.`;
            }
            return true;
        }
        if (name === 'matches') {
            let re: RegExp;
            if (arg instanceof RegExp) re = arg;
            else if (Array.isArray(arg) && arg[0] != null) {
                re = new RegExp(String(arg[0]), arg[1] != null ? String(arg[1]) : '');
            } else {
                re = new RegExp(String(arg));
            }
            if (!re.test(String(value))) {
                return (msg as string) || `${fieldKey} has an invalid format.`;
            }
            return true;
        }
        if (name === 'contains' && (typeof arg === 'string' || typeof arg === 'number')) {
            if (!String(value).includes(String(arg))) {
                return (msg as string) || `${fieldKey} must contain the required value.`;
            }
            return true;
        }
        if (name === 'notcontains' && (typeof arg === 'string' || typeof arg === 'number')) {
            if (String(value).includes(String(arg))) {
                return (msg as string) || `${fieldKey} may not contain the forbidden value.`;
            }
            return true;
        }
        if (name === 'isafter' || name === 'isbefore') {
            const a = toDate(value);
            const b = toDate(arg);
            if (!a) {
                return (msg as string) || `${fieldKey} is not a valid date.`;
            }
            if (b) {
                if (name === 'isafter' && !(a.getTime() > b.getTime())) {
                    return (msg as string) || `${fieldKey} is not after the limit.`;
                }
                if (name === 'isbefore' && !(a.getTime() < b.getTime())) {
                    return (msg as string) || `${fieldKey} is not before the limit.`;
                }
            }
            return true;
        }
        if (name === 'equals') {
            if (value !== arg) {
                return (msg as string) || `${fieldKey} has an invalid value.`;
            }
            return true;
        }
        return true;
    };
}

/**
 * O subvalidador (positivo) passou a true ⇒ `not` falha. Falha do sub (string/false) ⇒ `not` ok.
 */
function runNotValidator(spec: any, value: any, fieldKey: string, msg: string | undefined): boolean | string {
    const inner = spec?.not;
    if (typeof inner === 'function') {
        const r = inner(value) as any;
        if (r === true) {
            return msg || `${fieldKey} is invalid.`;
        }
        return true;
    }
    if (inner && typeof inner === 'object' && !Array.isArray(inner)) {
        for (const [k, v] of Object.entries(inner as Record<string, any>)) {
            if (isBuiltInValidatorKey(k) && typeof v !== 'function') {
                const r = buildBuiltinValidator(k, v, fieldKey)(value) as any;
                if (r === true) {
                    return msg || `${fieldKey} is invalid.`;
                }
                if (typeof r === 'string' || r === false) {
                    continue;
                }
            } else if (typeof v === 'function') {
                const r = (v as ValidatorFn)(value) as any;
                if (r === true) {
                    return msg || `${fieldKey} is invalid.`;
                }
            }
        }
    }
    return true;
}

/**
 * Aplica um par do objeto `validate`: built-in, função, ou `not: { ... }`.
 */
export async function runValidateEntry(
    key: string,
    validatorName: string,
    spec: unknown,
    value: any,
    _column: ColumnOptions
): Promise<boolean | string> {
    if (validatorName.toLowerCase() === 'not' && spec != null && typeof spec === 'object' && 'not' in (spec as any)) {
        return runNotValidator(spec, value, key, (spec as any).msg);
    }
    if (isBuiltInValidatorKey(validatorName) && typeof spec !== 'function') {
        return buildBuiltinValidator(validatorName, spec, key)(value) as any;
    }
    if (typeof spec === 'function') {
        return (await (spec as ValidatorFn)(value)) as any;
    }
    return `${key}: invalid validate entry "${validatorName}" (expected built-in or function)`;
}

File: src/orm_py/validators.py
from __future__ import annotations

import re
from datetime import datetime
from typing import Any, Callable, Mapping
from urllib.parse import urlparse
from uuid import UUID

ValidatorFn = Callable[[Any], bool | str]

BUILTIN_VALIDATOR_NAMES = {
    "notempty",
    "isnull",
    "isin",
    "notin",
    "min",
    "max",
    "len",
    "isemail",
    "isurl",
    "isip",
    "isuuid",
    "isdate",
    "isint",
    "isfloat",
    "isdecimal",
    "isnumeric",
    "isalpha",
    "isalphanumeric",
    "matches",
    "contains",
    "notcontains",
    "isafter",
    "isbefore",
    "equals",
    "not",
}

RE_EMAIL = re.compile(r"^[a-z0-9._%+\-]+@[a-z0-9.\-]+\.[a-z]{2,63}$", re.I)
RE_INT = re.compile(r"^-?\d+$")
RE_NUMERIC = re.compile(r"^-?\d+(\.\d+)?([eE][+\-]?\d+)?$")
RE_ALPHA = re.compile(r"^[A-Za-zÀ-ÿ\u0100-\uFFFF]+$")
RE_ALNUM = re.compile(r"^[0-9A-Za-zÀ-ÿ\u0100-\uFFFF]+$")
RE_IPV4 = re.compile(r"^(?:(?:25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)\.){3}(?:25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)$")


def is_builtin_validator_key(name: str) -> bool:
    return str(name).lower() in BUILTIN_VALIDATOR_NAMES


def _unpack(spec: Any) -> tuple[Any, str | None]:
    if isinstance(spec, Mapping) and ("args" in spec or "msg" in spec):
        return spec.get("args"), spec.get("msg")
    return spec, None


def _to_date(value: Any) -> datetime | None:
    if isinstance(value, datetime):
        return value
    if value is None:
        return None
    try:
        if isinstance(value, (int, float)):
            return datetime.fromtimestamp(value)
        return datetime.fromisoformat(str(value).replace("Z", "+00:00"))
    except Exception:
        return None


def _is_blank_value(value: Any) -> bool:
    if value is None:
        return True
    if isinstance(value, str) and value.strip() == "":
        return True
    if isinstance(value, (list, tuple, set)) and len(value) == 0:
        return True
    return False


def _parse_allowed_list(value: Any) -> list[Any]:
    if value is None:
        return []
    if isinstance(value, (list, tuple)):
        if len(value) > 0 and isinstance(value[0], (list, tuple)):
            return list(value[0])
        return list(value)
    return [value]


def run_not_validator(spec: Any, value: Any, field_key: str, msg: str | None = None) -> bool | str:
    inner = spec.get("not") if isinstance(spec, Mapping) else None
    if callable(inner):
        result = inner(value)
        if result is True:
            return msg or f"{field_key} is invalid."
        return True
    if isinstance(inner, Mapping):
        for key, entry in inner.items():
            if is_builtin_validator_key(str(key)) and not callable(entry):
                result = build_builtin_validator(str(key), entry, field_key)(value)
                if result is True:
                    return msg or f"{field_key} is invalid."
                continue
            if callable(entry):
                result = entry(value)
                if result is True:
                    return msg or f"{field_key} is invalid."
    return True


def build_builtin_validator(validator_name: str, spec: Any, field_key: str) -> ValidatorFn:
    name = validator_name.lower()
    arg, msg = _unpack(spec)

    if name == "not":
        return lambda value: run_not_validator(spec, value, field_key, msg)

    def _validator(value: Any) -> bool | str:
        if name == "notempty":
            if _is_blank_value(value):
                return msg or f"{field_key} cannot be empty."
            return True
        if name == "isnull":
            if value is not None:
                return msg or f"{field_key} must be null."
            return True
        if name in ("isin", "notin"):
            if value is None:
                return True
            options = _parse_allowed_list(arg)
            hit = any((x == value) for x in options)
            if name == "isin" and not hit:
                return msg or f"{field_key} is not a valid choice."
            if name == "notin" and hit:
                return msg or f"{field_key} is not an allowed value."
            return True
        if name in ("min", "max"):
            if value is None:
                return True
            try:
                n = float(arg)
            except Exception:
                return True
            if name == "min":
                if isinstance(value, (int, float)) and float(value) < n:
                    return msg or f"{field_key} must be >= {n:g}."
                if isinstance(value, str) and len(value) < int(n):
                    return msg or f"{field_key} is too short (min {int(n)})."
                if isinstance(value, (list, tuple, set)) and len(value) < int(n):
                    return msg or f"{field_key} has too few elements."
            if name == "max":
                if isinstance(value, (int, float)) and float(value) > n:
                    return msg or f"{field_key} must be <= {n:g}."
                if isinstance(value, str) and len(value) > int(n):
                    return msg or f"{field_key} is too long (max {int(n)})."
                if isinstance(value, (list, tuple, set)) and len(value) > int(n):
                    return msg or f"{field_key} has too many elements."
            return True
        if name == "len":
            if value is None:
                return True
            pair = arg if isinstance(arg, (list, tuple)) and len(arg) >= 2 else None
            if not pair:
                return True
            lo, hi = int(pair[0]), int(pair[1])
            if isinstance(value, (str, list, tuple, set)):
                size = len(value)
                if size < lo or size > hi:
                    return msg or f"{field_key} must have between {lo} and {hi} items."
            return True
        if value is None:
            return True
        if name == "isemail":
            if not RE_EMAIL.search(str(value)):
                return msg or f"{field_key} must be a valid email."
            return True
        if name == "isurl":
            parsed = urlparse(str(value))
            if not parsed.scheme or not parsed.netloc:
                return msg or f"{field_key} must be a valid URL."
            return True
        if name == "isip":
            if not RE_IPV4.search(str(value)):
                return msg or f"{field_key} must be a valid IPv4."
            return True
        if name == "isuuid":
            try:
                uid = UUID(str(value))
                if arg == 4 and uid.version != 4:
                    return msg or f"{field_key} must be a valid UUID v4."
            except Exception:
                return msg or f"{field_key} must be a valid UUID."
            return True
        if name == "isdate":
            if _to_date(value) is None:
                return msg or f"{field_key} must be a valid date."
            return True
        if name == "isint":
            if not RE_INT.search(str(value)):
                return msg or f"{field_key} must be an integer."
            return True
        if name in ("isfloat", "isdecimal", "isnumeric"):
            if not RE_NUMERIC.search(str(value)):
                return msg or f"{field_key} must be numeric."
            return True
        if name == "isalpha":
            if not RE_ALPHA.search(str(value)):
                return msg or f"{field_key} may only contain letters."
            return True
        if name == "isalphanumeric":
            if not RE_ALNUM.search(str(value)):
                return msg or f"{field_key} may only contain letters and numbers."
            return True
        if name == "matches":
            regex = None
            if isinstance(arg, re.Pattern):
                regex = arg
            elif isinstance(arg, (list, tuple)) and len(arg) > 0:
                regex = re.compile(str(arg[0]), str(arg[1]) if len(arg) > 1 else "")
            elif arg is not None:
                regex = re.compile(str(arg))
            if regex and not regex.search(str(value)):
                return msg or f"{field_key} has an invalid format."
            return True
        if name == "contains":
            if str(arg) not in str(value):
                return msg or f"{field_key} must contain the required value."
            return True
        if name == "notcontains":
            if str(arg) in str(value):
                return msg or f"{field_key} may not contain the forbidden value."
            return True
        if name in ("isafter", "isbefore"):
            a = _to_date(value)
            b = _to_date(arg)
            if a is None:
                return msg or f"{field_key} is not a valid date."
            if b is not None:
                if name == "isafter" and not (a > b):
                    return msg or f"{field_key} is not after the limit."
                if name == "isbefore" and not (a < b):
                    return msg or f"{field_key} is not before the limit."
            return True
        if name == "equals":
            if value != arg:
                return msg or f"{field_key} has an invalid value."
            return True
        return True

    return _validator


def run_validate_entry(key: str, validator_name: str, spec: Any, value: Any, _column: Mapping[str, Any]) -> bool | str:
    if validator_name.lower() == "not" and isinstance(spec, Mapping) and "not" in spec:
        return run_not_validator(spec, value, key, spec.get("msg"))
    if is_builtin_validator_key(validator_name) and not callable(spec):
        return build_builtin_validator(validator_name, spec, key)(value)
    if callable(spec):
        return spec(value)
    return f'{key}: invalid validate entry "{validator_name}" (expected built-in or function)'

